pkg/kubelet/kubelet_network_linux.go

```diff
@@ -95,27 +97,37 @@ func (kl *Kubelet) syncNetworkUtil(iptClient utiliptables.Interface) bool {
 		return false
 	}
 
+	if _, err := iptClient.EnsureChain(utiliptables.TableFilter, KubeFirewallInChain); err != nil {
+		klog.Errorf("Failed to ensure that %s chain %s exists: %v", utiliptables.TableFilter, KubeFirewallInChain, err)
+		return false
+	}
+
 	// drop all non-local packets to localhost if they're not part of an existing
 	// forwarded connection. See #90259
 	if !iptClient.IsIPv6() { // ipv6 doesn't have this issue
-		if _, err := iptClient.EnsureRule(utiliptables.Append, utiliptables.TableFilter, KubeFirewallChain,
+		if _, err := iptClient.EnsureRule(utiliptables.Append, utiliptables.TableFilter, KubeFirewallInChain,
 			"-m", "comment", "--comment", "block incoming localnet connections",
 			"--dst", "127.0.0.0/8",
-			"!", "--src", "127.0.0.0/8",
+			"!", "-i", "lo",
 			"-m", "conntrack",
 			"!", "--ctstate", "RELATED,ESTABLISHED,DNAT",
 			"-j", "DROP"); err != nil {
-			klog.Errorf("Failed to ensure rule to drop invalid localhost packets in %v chain %v: %v", utiliptables.TableFilter, KubeFirewallChain, err)
+			klog.Errorf("Failed to ensure rule to drop invalid localhost packets in %v chain %v: %v", utiliptables.TableFilter, KubeFirewallInChain, err)
 			return false
 		}
 	}
 
+	// Add KUBE-FIREWALL to OUTPUT, add KUBE-FIREWALL-IN to INPUT, add KUBE-FIREWALL to KUBE-FIREWALL-IN
 	if _, err := iptClient.EnsureRule(utiliptables.Prepend, utiliptables.TableFilter, utiliptables.ChainOutput, "-j", string(KubeFirewallChain)); err != nil {
 		klog.Errorf("Failed to ensure that %s chain %s jumps to %s: %v", utiliptables.TableFilter, utiliptables.ChainOutput, KubeFirewallChain, err)
 		return false
 	}
-	if _, err := iptClient.EnsureRule(utiliptables.Prepend, utiliptables.TableFilter, utiliptables.ChainInput, "-j", string(KubeFirewallChain)); err != nil {
-		klog.Errorf("Failed to ensure that %s chain %s jumps to %s: %v", utiliptables.TableFilter, utiliptables.ChainInput, KubeFirewallChain, err)
+	if _, err := iptClient.EnsureRule(utiliptables.Append, utiliptables.TableFilter, KubeFirewallInChain, "-j", string(KubeFirewallChain)); err != nil {
+		klog.Errorf("Failed to ensure that %s chain %s jumps to %s: %v", utiliptables.TableFilter, KubeFirewallInChain, KubeFirewallChain, err)
+		return false
+	}
+	if _, err := iptClient.EnsureRule(utiliptables.Prepend, utiliptables.TableFilter, utiliptables.ChainInput, "-j", string(KubeFirewallInChain)); err != nil {
+		klog.Errorf("Failed to ensure that %s chain %s jumps to %s: %v", utiliptables.TableFilter, utiliptables.ChainInput, KubeFirewallInChain, err)
```

Comment:
If I read correctly
```
- iptables -t filter -I INPUT -j KUBE-FIREWALL
+ iptables -t filter -A KUBE-FIREWALL-IN -j KUBE-FIREWALL
+ iptables -t filter -I INPUT -j KUBE-FIREWALL-IN
```
This works for me.
I'm curious why you didn't went with
```
  iptables -t filter -I INPUT -j KUBE-FIREWALL
+ iptables -t filter -I INPUT -j KUBE-FIREWALL-IN
```
(performance ? usability ?)

Reply:
Also what happens on upgrade ?

Reply:
Oh, upgrade! Good catch.

It was a while ago, but I was thinking that one would theoretically like to add `-j ALLOW` rules in `KUBE-FIREWALL-IN`, but that doesn't actually make any sense. I'll do what you suggest.

Reply:
updated.